[Issue #52] Add schema content + presetNodes ModelObject constructor

gahag opened the pull request:
# Adds a new constructor to ModelObject allowing the use of schema content with preset nodes and nodes to build.

## Description

As requested in the issue, there should be a constructor allowing prefetched schemas.

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
#52

## Motivation and Context

Discussed in the issue.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
As I've not found tests for the others constructors involving a resource location, i have not figured out a way to test this overload. Any tips on this?

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

gahag committed:
[Issue #52] Add schema content + presetNodes ModelObject constructor

tedcasey commented on utilities/ride-model-util/src/main/java/com/adobe/ride/utilities/model/ModelObject.java:
```diff
@@ -278,6 +278,43 @@ private void initializeModelObject(String resourceLocation, JSONObject presetNod
+        useRequiredOnly);
+  }
+
+  private void initializeModelObject(String resourceLocation, String modelString,
```
I don't know that it makes sense to break this out into a sub method.  Might as well put it in the constructor.

gahag replied:
I did it for the other constructors in the class did the same. I'll inline it right away.

tedcasey commented on utilities/ride-model-util/src/main/java/com/adobe/ride/utilities/model/ModelObject.java:
```diff
@@ -278,6 +278,43 @@ private void initializeModelObject(String resourceLocation, JSONObject presetNod
+   * @param useRequiredOnly boolean to determine whether to build only the required json nodes as
+   *        defined in the spec
+   */
+  public ModelObject(String resourceLocation, String modelString, JSONObject presetNodes,
```
Don't need resourceLocation here.  That is a reference to a json schema file on disk.  May want to take a look at this constructor as well: https://github.com/adobe/ride/blob/develop/utilities/ride-model-util/src/main/java/com/adobe/ride/utilities/model/ModelObject.java#L288

gahag replied:
I did it for there is already a constructor with the (String, JsonObject, Set<String>, boolean) signature. I had to find a new overload to avoid conflict, and this was the best compromise I've found that didn't require new types (as in wrapping the modelString in a new class). What do you suggest? Thanks for the reviews by the way.

tedcasey replied:
Hmm, I don't see the existing signature that you describe.  Can you link me to it in the code?

gahag replied:
Here it is: 
https://github.com/adobe/ride/blob/89c40d009b95e94e8d46790004629cae41392299/utilities/ride-model-util/src/main/java/com/adobe/ride/utilities/model/ModelObject.java#L260

tedcasey replied:
Ah, yes.  If you remove that property, it does collide.  Drat.

I guess we'll have to go with a static constructor.  Something like "createFromSchemaString".  Eventually, these constructors may need some refactoring, but a static constructor can fill this need in the near term.

gahag replied:
Sure, will do.

gahag replied:
Done!

tedcasey reviewed (changes requested):
Thanks for taking this on! 

Please take a look at the comments.

gahag committed:
Inline method `ModelObject::initializeModelObject`

gahag committed:
Remove unnecessary parameter from constructor, converting to static

Also, the constructor was converted to a static constructor to prevent overload collision.

tedcasey commented on utilities/ride-model-util/src/main/java/com/adobe/ride/utilities/model/ModelObject.java:
```diff
@@ -278,6 +278,37 @@ private void initializeModelObject(String resourceLocation, JSONObject presetNod
+   * with. Also only build a targeted set of nodes. The specified resource location is only used as
+   * metadata, so the content will not be fetched from it.
+   * 
+   * @param resourceLocation location of the schema within the project resources.
```
Please remove this parameter annotation.

gahag replied:
Oh, sure. Sorry for missing this out.

tedcasey reviewed (changes requested):
A few changes requested and one question:  

Can you attempt to test this change by loading one of the sample schemas as a string and instantiate with this method?

Thanks, again!

tedcasey commented on utilities/ride-model-util/src/main/java/com/adobe/ride/utilities/model/ModelObject.java:
```diff
@@ -278,6 +278,37 @@ private void initializeModelObject(String resourceLocation, JSONObject presetNod
+  /**
+   * Create a model object from a json schema string and use a predefined set of nodes to start
+   * with. Also only build a targeted set of nodes. The specified resource location is only used as
+   * metadata, so the content will not be fetched from it.
```
Can we tighten up this description a little?  Maybe something like: "Static constructor to initialize with a JSON schema string.  Optional methods for adding preset nodes, defining target nodes, and whether to use only required nodes."

gahag replied:
Sure!

tedcasey commented on utilities/ride-model-util/src/main/java/com/adobe/ride/utilities/model/ModelObject.java:
```diff
@@ -278,6 +278,37 @@ private void initializeModelObject(String resourceLocation, JSONObject presetNod
+   * @return ModelObject the model object
+   */
+  public static ModelObject createFromSchemaString(String modelString, JSONObject presetNodes,
+      Set<String> nodesToBuild, boolean useRequiredOnly) {
```
Can you add a null check for useRequiredOnly and set it to false if null?

gahag replied:
Null is only possible for object types, like `Boolean`, but this parameter has primitive type `boolean`, so I believe it can't be null. But the concern is valid, and it raised my attention to the `modelString` parameter. The other constructor accepting the modelString does not check if it's null, nor does mine. Should we do something about it?

tedcasey replied:
Good point.  Per the java docs, unset booleans default to false, which works for our purposes. However, we should honor the argument.  Please set the global value as indicated by the argument.  The other constructor uses the internal method for this, which, while a bit superfluous, works.  We should continue to use it to be consistent: https://github.com/adobe/ride/blob/develop/utilities/ride-model-util/src/main/java/com/adobe/ride/utilities/model/ModelObject.java#L289

gahag replied:
As I'm already calling the other constructor with the useRequiredOnly argument, the property is already being set. Is it enough?

tedcasey replied:
ah, yes.  Right.  That is sufficient.

gahag committed:
Improve javadoc for ModelObject::createFromSchemaString

gahag committed:
Fix test ModelObjectTest::testStringJsonSchema

gahag commented:
I'm writing the test case for my constructor, and I've stumbled on two issues. The first one is that an old test was not passing. I've managed to fix it on my last commit. The second issue is that the `ModelObject(String modelString, boolean useRequiredOnly)` constructor apparently leaves the object in an invalid state. I'm still working to see if there's something I can do. Any hints on that will be appreciated.

gahag committed:
Add ModelObjectTest::testSchemaStringStatic test

tedcasey commented on utilities/ride-model-util/src/test/java/com/adobe/ride/utilities/model/test/ModelObjectTest.java:
```diff
@@ -89,7 +89,7 @@ public void testStringJsonSchema() throws ModelSearchException {
     try {
       InputStream inputStream =
-          ClassLoader.class.getResourceAsStream("/schemas/TestService/simple.json");
+          ModelObject.class.getResourceAsStream("/schemas/TestService/simple.json");
```
Don't change this one.  ClassLoader is correct here.

gahag replied:
I've replace it with the manner used in the ModelObject class to fetch the schema string, and it works. As I've commented below, the develop branch fails this test due to a null pointer in this exact line. What should we do?

tedcasey replied:
Hmm, strange.  Let me check this out.

gahag replied:
Any news on this?

tedcasey replied:
Sorry, got tied up with another pressing issue.  Will have a response a little later today.

tedcasey replied:
So, I've run verify on the current develop branch and everything passes.

This is the command I used: mvn clean verify -Dgpg.skip

Tests run: 14, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.763 sec

Can you show me the command you are using to run your tests?

gahag replied:
I'm using `mvn -f build/pom.xml test`. I've tried now with `mvn -f build/pom.xml -Dgpg.skip verify` as you've recommended, and the test still fails. I'm on the origin/develop branch, up to date, working dir clean. Maybe it has something to do with my java version? I'm using java 13 on arch linux.

gahag replied:
Found it: the tests pass with java 8. Apparently there is a breaking change to java 13. What should we do?

gahag replied:
I'm downloading java 10 to test whether it works.

gahag replied:
The test fails with java 10.

gahag replied:
So, apparently this is our issue:
https://stackoverflow.com/questions/47613602/java-9-0-classloadergetresourceasstream-nullpointerexception

gahag replied:
I'll revert this change and keep the java 8 compatible version. The java 9+ compatibility issue can be addressed in another thread.

gahag replied:
Done, I believe this topic is settled now.

tedcasey commented:
Does that ModelObject constructor work outside of your static constructor?  I would test that first.

Can you post your stack error, if you're getting one, or any additional information about how you know the instance is in an invalid state?

Can you post your test code?

gahag commented:
I've checked out the develop branch from the main repo to run the test, and the test fails:
```
Tests run: 14, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.565 sec <<< FAILURE!
testStringJsonSchema(com.adobe.ride.utilities.model.test.ModelObjectTest)  Time elapsed: 0.003 sec  <<< FAILURE!
java.lang.NullPointerException
        at java.base/java.io.Reader.<init>(Reader.java:167)
        at java.base/java.io.InputStreamReader.<init>(InputStreamReader.java:113)
        at org.apache.commons.io.IOUtils.copy(IOUtils.java:2272)
        at org.apache.commons.io.IOUtils.toString(IOUtils.java:1041)
        at com.adobe.ride.utilities.model.test.ModelObjectTest.testStringJsonSchema(ModelObjectTest.java:94)
```

gahag commented:
In my last commit (d20318b) I've managed to get my test working.

gahag committed:
Revert "Fix test ModelObjectTest::testStringJsonSchema"

The test was actually correct under java 8. A breaking change introduced in java 9 broke
the test.

This reverts commit 1945a0d438a9f4fdf45ca5a0b1ee7e07eadd1118.

gahag commented:
I've reverted the change to the existing test (see my comments in the resolved section above for details.), and added a new test for the code I've introduced. Everything seems correct now, so tell me if there's still anything to be improved. Thanks for your attention.

tedcasey commented:
Thanks for running this down!  Can you do me a favor a write up a new issue regarding the Java 10+ issue?  We'll need to address it at some point.

Taking a final look now!

tedcasey commented on utilities/ride-model-util/src/test/java/com/adobe/ride/utilities/model/test/ModelObjectTest.java:
```diff
@@ -106,6 +106,46 @@ public void testStringJsonSchema() throws ModelSearchException {
+
+    // Targets
+    Set<String> targetNodes = new HashSet<String>();
+    targetNodes.add("/title");
```
Thanks for adding this test!  One more final request.  Can you add a few more target nodes here to ensure that things other than preset nodes are added to the instance?

gahag replied:
Done!

tedcasey reviewed (commented):
Once one of more additional target nodes are added, the PR is approved.

gahag committed:
Add targetNode to ModelObjectTest::testSchemaStringStatic

Also switch schema from simple.json to entitylink.json.

tedcasey reviewed (approved):
Thanks for your contribution!

tedcasey commented:
Now that you have an approved PR, you're welcome to have a Ride T-shirt, if you like.  I'll just need to know where to send it.  You can send me the address through the linked in link on the readme (so you don't have to publicly broadcast it).

If not, that's ok too.

Thanks again for contributing!  Looking forward to more ;)

gahag commented:
I can't seem to find the link for informing the address. Where is it?

tedcasey commented:
Hello.  Sorry for the delay.  Here is it: https://www.linkedin.com/in/ted-casey-0074064/

Also, please let me know a size when you ping.  S/M/L is avail (don't know if you needed larger, but I don't have any XL currently - might order more if there is demand).

Thanks again!
